protocols/kad/src/query.rs

```diff
@@ -308,13 +341,38 @@ impl<TInner> Query<TInner> {
         state
     }
 
+    /// Tries to (gracefully) finish the query prematurely, providing the peers
+    /// that are no longer of interest for further progress of the query.
+    ///
+    /// A query may require that in order to finish gracefully a certain subset
+    /// of peers must be contacted. E.g. in the case of disjoint query paths a
+    /// query may only finish gracefully if every path contacted a peer whose
+    /// response permits termination of the query. The given peers are those for
+    /// which this is considered to be the case, i.e. for which a termination
+    /// condition is satisfied.
+    ///
+    /// A finished query immediately stops yielding new peers to contact and
+    /// will be reported by [`QueryPool::poll`] via
+    /// [`QueryPoolState::Finished`].
+    pub fn try_finish<I>(&mut self, peers: I)
```

Comment:
> I am not in favor of altering the function signature for a debug statement. Wouldn't that confuse more than it helps?

I tend to view the module APIs also in isolation and when just looking at `try_finish()` the first thing I wonder is how I am supposed to know about the outcome of this attempt, i.e. is the query now finished or not? Of course we already have `is_finished()` and I suppose we could use that as well, but indicating success/failure from `try_finish()` itself via a return value seems a bit more straight-forward. Just because, _right now_, the only place where we would use this return value is so that we can add some useful logging, because for the program flow at that point we don't actually care whether the query did indeed finish, does not make the fact that `try_finish` returns success/failure less valuable, in my opinion.

> What do you think of b69f24faf8419768d470c73bdbed24c5a11e43c7 instead?

I don't think these logs are very useful because they lack too much context, like the query ID or the peer ID of the latest response, both of which are available at the call-site of `try_finish()`.